File: statistical_analysis_spr2025.m
%% Statistical Analysis of EEG Features in Relation to Genetic Diagnosis
% This script performs statistical analysis on EEG features including:
% 1. Data preprocessing (normalization, handling missing values)
% 2. Multiple regression analysis with covariates (age, sex)
% 3. Statistical testing and results table generation

%% Set up paths
root_dir = '/Users/emmanuelle.coutu-nadeau/Library/Mobile Documents/com~apple~CloudDocs/UdeM/MSc Psycho/LABO NED - Personal Drive/Code/GENiAL';
preprocessed_data_path = fullfile(root_dir, 'Data', 'Final', 'GENIAL-DB-preprocessed-V2.csv');
output_dir = fullfile(root_dir, 'Results', 'Statistical_Analysis');

% Create output directory if it doesn't exist
if ~exist(output_dir, 'dir')
    mkdir(output_dir);
end

%% Load and prepare data
fprintf('Loading and preparing data...\n');

% Load the preprocessed data
data = readtable(preprocessed_data_path);

% Identify EEG feature columns
column_names = data.Properties.VariableNames;
eeg_cols = startsWith(column_names, 'EEG_');
non_numeric_cols = {'EEG_attempted', 'EEG_site', 'EEG_date', 'EEG_age', 'EEG_Age', 'EEG_Sex'};
eeg_cols = eeg_cols & ~ismember(column_names, non_numeric_cols);
eeg_feature_names = column_names(eeg_cols);

% Convert sex to binary
sex_binary = double(strcmp(data.Sex_at_birth, 'Male'));

% Create diagnostic groups
% 0: Control (diag_control = 1)
% 1: Neurodev only (diag_neurodev = 1 and diag_genetic_carrier = 0)
% 2: Genetic carrier (diag_genetic_carrier = 1)
diagnostic_group = zeros(height(data), 1);
diagnostic_group(data.diag_neurodev == 1 & data.diag_genetic_carrier == 0) = 1;
diagnostic_group(data.diag_genetic_carrier == 1) = 2;

% Print group sizes
fprintf('\nDiagnostic group sizes:\n');
fprintf('Controls: %d\n', sum(diagnostic_group == 0));
fprintf('Neurodev only: %d\n', sum(diagnostic_group == 1));
fprintf('Genetic carriers: %d\n', sum(diagnostic_group == 2));

%% Normalize data
fprintf('\nNormalizing data...\n');

% Extract EEG features
eeg_features = table2array(data(:, eeg_cols));

% Report NaN statistics
fprintf('\nMissing data statistics:\n');
nan_counts = sum(isnan(eeg_features), 1);
total_samples = size(eeg_features, 1);
for i = 1:length(eeg_feature_names)
    fprintf('%s: %d/%d (%.1f%%) missing values\n', ...
        eeg_feature_names{i}, nan_counts(i), total_samples, ...
        nan_counts(i)/total_samples*100);
end

% Z-score normalization (handling NaN values)
eeg_features_norm = zeros(size(eeg_features));
for i = 1:size(eeg_features, 2)
    valid_idx = ~isnan(eeg_features(:, i));
    if sum(valid_idx) > 0  % Only normalize if there are valid values
        eeg_features_norm(valid_idx, i) = zscore(eeg_features(valid_idx, i));
    end
end

%% Run multiple regression analysis
fprintf('\nRunning multiple regression analysis...\n');

% Initialize results structure
n_features = length(eeg_feature_names);
results = struct('Feature', cell(n_features, 1), ...
                'R_squared', [], ...
                'Adj_R_squared', [], ...
                'F_statistic', [], ...
                'F_pvalue', [], ...
                'Neurodev_coef', [], ...
                'Neurodev_pvalue', [], ...
                'Genetic_coef', [], ...
                'Genetic_pvalue', [], ...
                'Age_coef', [], ...
                'Age_pvalue', [], ...
                'Sex_coef', [], ...
                'Sex_pvalue', [], ...
                'N_samples', []);  % Add field for number of samples used

% Create design matrix
X = [ones(height(data), 1), data.EEG_age, sex_binary, ...
     diagnostic_group == 1, diagnostic_group == 2];

% Check for multicollinearity
corr_matrix = corr(X(:,2:end)); % Exclude intercept
if any(abs(corr_matrix(tril(ones(size(corr_matrix)),-1)==1)) > 0.9)
    warning('High multicollinearity detected in predictors. Results may be unreliable.');
end

% Process all features, handling missing values case by case
for i = 1:n_features
    % Get current feature
    y = eeg_features_norm(:, i);
    valid_idx = ~isnan(y);
    
    % Skip if not enough valid samples
    if sum(valid_idx) < size(X, 2)
        fprintf('Warning: Not enough valid samples for feature %s (%d/%d). Skipping.\n', ...
            eeg_feature_names{i}, sum(valid_idx), size(X, 2));
        continue;
    end
    
    y_valid = y(valid_idx);
    X_valid = X(valid_idx, :);
    
    % Fit model
    [b, bint, r, rint, stats] = regress(y_valid, X_valid);
    
    % Store results
    results(i).Feature = eeg_feature_names{i};
    results(i).N_samples = sum(valid_idx);
    results(i).R_squared = stats(1);
    results(i).F_statistic = stats(2);
    results(i).F_pvalue = stats(3);
    results(i).Neurodev_coef = b(4);
    results(i).Genetic_coef = b(5);
    results(i).Age_coef = b(2);
    results(i).Sex_coef = b(3);
    
    % Calculate p-values
    df = length(y_valid) - length(b);
    t_stats = b ./ sqrt(diag(inv(X_valid'*X_valid))*stats(4));
    p_values = 2 * (1 - tcdf(abs(t_stats), df));
    
    results(i).Neurodev_pvalue = p_values(4);
    results(i).Genetic_pvalue = p_values(5);
    results(i).Age_pvalue = p_values(2);
    results(i).Sex_pvalue = p_values(3);
end

%% Apply FDR correction
% Implement Benjamini-Hochberg FDR correction
function p_adj = bh_fdr(p)
    m = length(p);
    [~, idx] = sort(p);
    p_adj = zeros(size(p));
    p_adj(idx) = min(1, p(idx) .* m ./ (1:m)');
end

neurodev_pvals = [results.Neurodev_pvalue];
genetic_pvals = [results.Genetic_pvalue];

% Apply FDR correction
neurodev_pvals_adj = bh_fdr(neurodev_pvals);
genetic_pvals_adj = bh_fdr(genetic_pvals);

for i = 1:n_features
    results(i).Neurodev_pvalue_adj = neurodev_pvals_adj(i);
    results(i).Genetic_pvalue_adj = genetic_pvals_adj(i);
end

%% Save results
fprintf('\nSaving results...\n');

% Convert results structure to table
results_table = struct2table(results);

% Save to CSV
writetable(results_table, fullfile(output_dir, 'regression_results.csv'));

% Create and save summary report
fid = fopen(fullfile(output_dir, 'analysis_summary.txt'), 'w');
fprintf(fid, 'Statistical Analysis Summary\n');
fprintf(fid, '==========================\n\n');
fprintf(fid, 'Total features analyzed: %d\n', n_features);
fprintf(fid, 'Features significant for Neurodevelopmental Group: %d\n', ...
    sum([results.Neurodev_pvalue_adj] < 0.05));
fprintf(fid, 'Features significant for Genetic group: %d\n\n', ...
    sum([results.Genetic_pvalue_adj] < 0.05));

% Write significant features details
sig_neurodev = [results.Neurodev_pvalue_adj] < 0.05;
sig_genetic = [results.Genetic_pvalue_adj] < 0.05;

if any(sig_neurodev)
    fprintf(fid, '\nSignificant Features for Neurodevelopmental Group:\n');
    fprintf(fid, '--------------------------------------------\n');
    for i = find(sig_neurodev)
        fprintf(fid, '\nFeature: %s\n', results(i).Feature);
        fprintf(fid, 'Coefficient: %.4f\n', results(i).Neurodev_coef);
        fprintf(fid, 'P-value (FDR corrected): %.4e\n', results(i).Neurodev_pvalue_adj);
        fprintf(fid, 'R-squared: %.4f\n', results(i).R_squared);
        fprintf(fid, 'Covariates:\n');
        fprintf(fid, '  Age: coef = %.4f, p = %.4e\n', results(i).Age_coef, results(i).Age_pvalue);
        fprintf(fid, '  Sex: coef = %.4f, p = %.4e\n', results(i).Sex_coef, results(i).Sex_pvalue);
    end
end

if any(sig_genetic)
    fprintf(fid, '\nSignificant Features for Genetic Carrier Group:\n');
    fprintf(fid, '------------------------------------------\n');
    for i = find(sig_genetic)
        fprintf(fid, '\nFeature: %s\n', results(i).Feature);
        fprintf(fid, 'Coefficient: %.4f\n', results(i).Genetic_coef);
        fprintf(fid, 'P-value (FDR corrected): %.4e\n', results(i).Genetic_pvalue_adj);
        fprintf(fid, 'R-squared: %.4f\n', results(i).R_squared);
        fprintf(fid, 'Covariates:\n');
        fprintf(fid, '  Age: coef = %.4f, p = %.4e\n', results(i).Age_coef, results(i).Age_pvalue);
        fprintf(fid, '  Sex: coef = %.4f, p = %.4e\n', results(i).Sex_coef, results(i).Sex_pvalue);
    end
end

fclose(fid);

fprintf('\nAnalysis complete! Results saved in: %s\n', output_dir); 

File: Python scripts/statistical_analysis-spr2025.py
#!/usr/bin/env python3
"""
This script performs statistical analysis on the EEG features in relation to genetic diagnosis.
It includes:
1. Data preprocessing (normalization, handling missing values)
2. Multiple regression analysis with covariates (age, sex)
3. Statistical testing and visualization
"""

import os
import pandas as pd
import numpy as np
from scipy import stats
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from statsmodels.stats.multitest import multipletests
import statsmodels.api as sm

# Directory paths
root_dir = "/Users/emmanuelle.coutu-nadeau/Library/Mobile Documents/com~apple~CloudDocs/UdeM/MSc Psycho/LABO NED - Personal Drive/Code/GENiAL/"
preprocessed_data_path = os.path.join(root_dir, 'Data/Final/GENIAL-DB-preprocessed-V2.csv')
output_dir = os.path.join(root_dir, 'Results/Statistical_Analysis/')

# Create output directory if it doesn't exist
os.makedirs(output_dir, exist_ok=True)

def load_and_prepare_data():
    """Load and prepare the data for analysis."""
    print("Loading and preparing data...")
    
    # Load the preprocessed data
    df = pd.read_csv(preprocessed_data_path)
    
    # Identify EEG feature columns (excluding non-numeric columns)
    non_numeric_cols = ['EEG_attempted', 'EEG_site', 'EEG_date', 'EEG_age', 'EEG_Age', 'EEG_Sex']
    eeg_cols = [col for col in df.columns if col.startswith('EEG_') and col not in non_numeric_cols]
    
    # Convert sex to binary (assuming 'Sex_at_birth' is the column name)
    df['Sex_binary'] = (df['Sex_at_birth'] == 'Male').astype(int)
    
    # Ensure diagnostic columns are integers
    df['diag_control'] = df['diag_control'].astype(int)
    df['diag_neurodev'] = df['diag_neurodev'].astype(int)
    df['diag_genetic_carrier'] = df['diag_genetic_carrier'].astype(int)
    
    # Create diagnostic groups
    # 0: Control (diag_control = 1)
    # 1: Neurodev only (diag_neurodev = 1 and diag_genetic_carrier = 0)
    # 2: Genetic carrier (diag_genetic_carrier = 1)
    
    df['diagnostic_group'] = 0  # Default to control
    df.loc[(df['diag_neurodev'] == 1) & (df['diag_genetic_carrier'] == 0), 'diagnostic_group'] = 1  # Neurodev only
    df.loc[df['diag_genetic_carrier'] == 1, 'diagnostic_group'] = 2  # Genetic carrier
    
    # Print group sizes for verification
    print("\nDiagnostic group sizes:")
    print("Controls:", sum(df['diagnostic_group'] == 0))
    print("Neurodev only:", sum(df['diagnostic_group'] == 1))
    print("Genetic carriers:", sum(df['diagnostic_group'] == 2))

    # Convert EEG columns to numeric, dropping any that can't be converted
    numeric_eeg_cols = []
    for col in eeg_cols:
        try:
            df[col] = pd.to_numeric(df[col], errors='coerce')
            # Only keep columns that have at least some numeric values
            if df[col].notna().any():
                numeric_eeg_cols.append(col)
        except:
            print(f"Skipping non-numeric column: {col}")
    
    print(f"\nNumber of numeric EEG features for analysis: {len(numeric_eeg_cols)}")
    
    return df, numeric_eeg_cols

def normalize_data(df, feature_cols):
    """Normalize EEG features using z-score standardization."""
    print("Normalizing data...")
    
    scaler = StandardScaler()
    df_normalized = df.copy()
    df_normalized[feature_cols] = scaler.fit_transform(df[feature_cols])
    
    return df_normalized

def check_assumptions(model, X, y, feature_name):
    """Check regression assumptions and create diagnostic plots."""
    # Residuals
    residuals = model.resid
    fitted_values = model.fittedvalues
    
    # Create figure with subplots
    fig, axes = plt.subplots(2, 2, figsize=(15, 10))
    
    # 1. Residuals vs Fitted
    axes[0, 0].scatter(fitted_values, residuals)
    axes[0, 0].axhline(y=0, color='r', linestyle='--')
    axes[0, 0].set_xlabel('Fitted values')
    axes[0, 0].set_ylabel('Residuals')
    axes[0, 0].set_title('Residuals vs Fitted')
    
    # 2. Q-Q plot
    sm.graphics.qqplot(residuals, dist=stats.norm, line='45', fit=True, ax=axes[0, 1])
    axes[0, 1].set_title('Q-Q Plot')
    
    # 3. Scale-Location
    axes[1, 0].scatter(fitted_values, np.sqrt(np.abs(residuals)))
    axes[1, 0].set_xlabel('Fitted values')
    axes[1, 0].set_ylabel('√|Residuals|')
    axes[1, 0].set_title('Scale-Location')
    
    # 4. Leverage plot
    influence = model.get_influence()
    leverage = influence.hat_matrix_diag
    axes[1, 1].scatter(leverage, residuals)
    axes[1, 1].set_xlabel('Leverage')
    axes[1, 1].set_ylabel('Residuals')
    axes[1, 1].set_title('Residuals vs Leverage')
    
    plt.tight_layout()
    plt.savefig(os.path.join(output_dir, f'diagnostic_plots_{feature_name}.png'))
    plt.close()

def run_multiple_regression(df, eeg_cols):
    """Run multiple regression for each EEG feature."""
    print("Running multiple regression analysis...")
    
    results = []
    
    for feature in eeg_cols:
        # Create dummy variables for diagnostic groups (control is reference)
        diagnostic_dummies = pd.get_dummies(df['diagnostic_group'], prefix='group', drop_first=True)
        
        # Prepare the data with dummy variables
        X = pd.concat([
            df[['EEG_age', 'Sex_binary']],
            diagnostic_dummies
        ], axis=1)
        
        y = df[feature]
        
        # Add constant for statsmodels
        X = sm.add_constant(X)
        
        # Fit the model
        model = sm.OLS(y, X).fit()
        
        # Check assumptions
        check_assumptions(model, X, y, feature)
        
        # Store results
        results.append({
            'Feature': feature,
            'R_squared': model.rsquared,
            'Adj_R_squared': model.rsquared_adj,
            'F_statistic': model.fvalue,
            'F_pvalue': model.f_pvalue,
            'Neurodev_coef': model.params['group_1'],
            'Neurodev_pvalue': model.pvalues['group_1'],
            'Genetic_coef': model.params['group_2'],
            'Genetic_pvalue': model.pvalues['group_2'],
            'Age_coef': model.params['EEG_age'],
            'Age_pvalue': model.pvalues['EEG_age'],
            'Sex_coef': model.params['Sex_binary'],
            'Sex_pvalue': model.pvalues['Sex_binary']
        })
    
    # Convert results to DataFrame
    results_df = pd.DataFrame(results)
    
    # Apply FDR correction for multiple comparisons (separately for each group)
    _, results_df['Neurodev_pvalue_adj'] = multipletests(results_df['Neurodev_pvalue'], 
                                                        method='fdr_bh')
    _, results_df['Genetic_pvalue_adj'] = multipletests(results_df['Genetic_pvalue'], 
                                                       method='fdr_bh')
    
    return results_df

def create_summary_visualizations(results_df, df):
    """Create summary visualizations of the results."""
    print("Creating visualizations...")
    
    # 1. Significant features plot for both groups
    plt.figure(figsize=(15, 6))
    
    # Get significant features for either group
    sig_features_neurodev = results_df[results_df['Neurodev_pvalue_adj'] < 0.05]
    sig_features_genetic = results_df[results_df['Genetic_pvalue_adj'] < 0.05]
    sig_features = pd.concat([sig_features_neurodev, sig_features_genetic]).drop_duplicates()
    
    if not sig_features.empty:
        # Create a melted dataframe for plotting
        plot_data = pd.melt(sig_features,
                           id_vars=['Feature'],
                           value_vars=['Neurodev_coef', 'Genetic_coef'],
                           var_name='Group',
                           value_name='Coefficient')
        plot_data['Group'] = plot_data['Group'].map({
            'Neurodev_coef': 'Neurodev Only',
            'Genetic_coef': 'Genetic Carrier'
        })
        
        # Create grouped bar plot
        sns.barplot(data=plot_data,
                   x='Feature',
                   y='Coefficient',
                   hue='Group',
                   palette='viridis')
        plt.xticks(rotation=45, ha='right')
        plt.title('Significant EEG Features by Group (FDR corrected p < 0.05)')
        plt.tight_layout()
        plt.savefig(os.path.join(output_dir, 'significant_features.png'))
        plt.close()
    
    # 2. Correlation matrix of significant features
    if not sig_features.empty:
        sig_feature_names = sig_features['Feature'].tolist()
        corr_matrix = df[sig_feature_names].corr()
        
        plt.figure(figsize=(10, 8))
        sns.heatmap(corr_matrix, annot=True, cmap='coolwarm', center=0)
        plt.title('Correlation Matrix of Significant Features')
        plt.tight_layout()
        plt.savefig(os.path.join(output_dir, 'correlation_matrix.png'))
        plt.close()

def save_results(results_df):
    """Save the analysis results."""
    print("Saving results...")
    
    # Save detailed results
    results_df.to_csv(os.path.join(output_dir, 'regression_results.csv'), index=False)
    
    # Create summary of significant findings
    sig_neurodev = results_df[results_df['Neurodev_pvalue_adj'] < 0.05]
    sig_genetic = results_df[results_df['Genetic_pvalue_adj'] < 0.05]
    
    # Save summary report
    with open(os.path.join(output_dir, 'analysis_summary.txt'), 'w') as f:
        f.write("Statistical Analysis Summary\n")
        f.write("==========================\n\n")
        
        f.write(f"Total features analyzed: {len(results_df)}\n")
        f.write(f"Features significant for Neurodevelopmental Group: {len(sig_neurodev)}\n")
        f.write(f"Features significant for Genetic group: {len(sig_genetic)}\n\n")
        
        if not sig_neurodev.empty:
            f.write("\nSignificant Features for Neurodevelopmental Group:\n")
            f.write("--------------------------------------------\n")
            for _, row in sig_neurodev.iterrows():
                f.write(f"\nFeature: {row['Feature']}\n")
                f.write(f"Coefficient: {row['Neurodev_coef']:.4f}\n")
                f.write(f"P-value (FDR corrected): {row['Neurodev_pvalue_adj']:.4e}\n")
                f.write(f"R-squared: {row['R_squared']:.4f}\n")
                f.write("Covariates:\n")
                f.write(f"  Age: coef = {row['Age_coef']:.4f}, p = {row['Age_pvalue']:.4e}\n")
                f.write(f"  Sex: coef = {row['Sex_coef']:.4f}, p = {row['Sex_pvalue']:.4e}\n")
        
        if not sig_genetic.empty:
            f.write("\nSignificant Features for Genetic Carrier Group:\n")
            f.write("------------------------------------------\n")
            for _, row in sig_genetic.iterrows():
                f.write(f"\nFeature: {row['Feature']}\n")
                f.write(f"Coefficient: {row['Genetic_coef']:.4f}\n")
                f.write(f"P-value (FDR corrected): {row['Genetic_pvalue_adj']:.4e}\n")
                f.write(f"R-squared: {row['R_squared']:.4f}\n")
                f.write("Covariates:\n")
                f.write(f"  Age: coef = {row['Age_coef']:.4f}, p = {row['Age_pvalue']:.4e}\n")
                f.write(f"  Sex: coef = {row['Sex_coef']:.4f}, p = {row['Sex_pvalue']:.4e}\n")

def main():
    # Load and prepare data
    df, eeg_cols = load_and_prepare_data()
    
    # Normalize EEG features
    df_normalized = normalize_data(df, eeg_cols)
    
    # Run multiple regression analysis
    results_df = run_multiple_regression(df_normalized, eeg_cols)
    
    # Create visualizations
    create_summary_visualizations(results_df, df_normalized)
    
    # Save results
    save_results(results_df)
    
    print("Analysis complete! Results saved in:", output_dir)

if __name__ == "__main__":
    main() 